contracts/registry/src/registry.rs

```diff
@@ -317,4 +317,101 @@ impl SBTRegistry for Contract {
+            for token in tokens {
+                // update balances
+                let class_id = token.metadata.class;
+                let balance_key = &BalanceKey {
+                    issuer_id,
+                    owner: owner.clone(),
+                    class_id: token.metadata.class,
+                };
+                self.balances.remove(balance_key);
+
+                // collect the info about the tokens revoked per class
+                // to update the balance accordingly
+                revoked_per_class
+                    .entry(class_id)
+                    .and_modify(|key_value| *key_value += 1)
+                    .or_insert(1);
+
+                // remove from issuer_tokens
+                self.issuer_tokens.remove(&IssuerTokenId {
+                    issuer_id,
+                    token: token.token,
+                });
+                token_ids.push(token.token);
+            }
+
+            // update supply by owner
+            let old_supply = self
+                .supply_by_owner
+                .get(&(owner.clone(), issuer_id))
+                .unwrap();
+            self.supply_by_owner.insert(
+                &(owner.clone(), issuer_id),
+                &(old_supply - &(tokens.len() as u64)),
+            );
+
+            // update supply by class
+            for (class_id, tokens_revoked) in revoked_per_class {
+                let old_supply = self.supply_by_class.get(&(issuer_id, class_id)).unwrap();
+                self.supply_by_class
+                    .insert(&(issuer_id, class_id), &(&old_supply - &tokens_revoked));
+            }
+
+            // update supply by issuer
+            let supply_by_issuer = self.supply_by_issuer.get(&(issuer_id)).unwrap_or(0);
+            self.supply_by_issuer
+                .insert(&(issuer_id), &(supply_by_issuer - tokens_burned));
+
+            // emit event
+            SbtTokensEvent {
+                issuer: issuer.clone(),
+                tokens: token_ids.clone(),
+            }
+            .emit_burn();
+        } else {
+            let current_timestamp_ms = env::block_timestamp_ms();
+            // revoke
+            for mut token in tokens.clone() {
+                // update expire date for all tokens to current_timestamp
+                token.metadata.expires_at = Some(current_timestamp_ms);
+                let token_data = TokenData {
+                    owner: owner.clone(),
+                    metadata: VerTokenMetadata::V1(token.metadata),
+                };
+                self.issuer_tokens.insert(
+                    &IssuerTokenId {
+                        issuer_id,
+                        token: token.token,
+                    },
+                    &token_data,
+                );
+                token_ids.push(token.token);
+            }
+        }
+        SbtTokensEvent {
+            issuer,
+            tokens: token_ids,
+        }
+        .emit_revoke();
```

Comment:
This needs to be double checked and in case of changes updated: https://www.pivotaltracker.com/n/projects/2623274/stories/185413931

Reply:
This is correct